futures-util/src/future/abortable.rs

```diff
@@ -0,0 +1,131 @@
+use futures_core::future::Future;
+use futures_core::task::{self, AtomicWaker, Poll};
+use std::marker::Unpin;
+use std::mem::PinMut;
+use std::sync::Arc;
+use std::sync::atomic::{AtomicBool, Ordering};
+
+/// A future which can be remotely short-circuited using an `AbortHandle`.
+#[derive(Debug, Clone)]
+#[must_use = "futures do nothing unless polled"]
+pub struct Abortable<T> {
+    future: T,
+    inner: Arc<AbortInner>,
+}
+
+impl<T: Unpin> Unpin for Abortable<T> {}
+
+impl<T> Abortable<T> where T: Future {
+    /// Creates a new `Abortable` future using an existing `AbortRegistration`.
+    /// `AbortRegistration`s can be acquired through `AbortHandle::new`.
+    ///
+    /// When `abort` is called on the handle tied to `reg` or if `abort` has
+    /// already been called, the future will complete immediately without making
+    /// any further progress.
+    pub fn new(future: T, reg: AbortRegistration) -> Self {
+        Abortable {
+            future: future,
+            inner: reg.inner,
+        }
+    }
+
+    unsafe_pinned!(future: T);
+}
+
+/// A registration handle for a `Abortable` future.
+/// Values of this type can be acquired from `AbortHandle::new` and are used
+/// in calls to `Abortable::new`.
+#[derive(Debug, Clone)]
+pub struct AbortRegistration {
+    inner: Arc<AbortInner>,
+}
+
+/// A handle to a `Abortable` future.
+#[derive(Debug, Clone)]
+pub struct AbortHandle {
+    inner: Arc<AbortInner>,
+}
+
+impl AbortHandle {
+    /// Creates a new `AbortHandle` which can be used to abort a running future.
+    /// This function also returns a `AbortRegistration` which is used to
+    /// pair the handle with a `Abortable` future.
+    ///
+    /// This function is usually paired with a call to `Abortable::new`.
+    pub fn new() -> (Self, AbortRegistration) {
+        let inner = Arc::new(AbortInner {
+            waker: AtomicWaker::new(),
+            cancel: AtomicBool::new(false),
+        });
+
+        (
+            AbortHandle {
+                inner: inner.clone(),
+            },
+            AbortRegistration {
+                inner: inner,
+            },
+        )
+    }
+}
+
+// Inner type storing the waker to awaken and a bool indicating that it
+// should be cancelled.
+#[derive(Debug)]
+struct AbortInner {
+    waker: AtomicWaker,
+    cancel: AtomicBool,
+}
+
+/// Creates a new `Abortable` future and a `AbortHandle` which can be used to stop it.
+///
+/// This function is a convenient (but less flexible) alternative to calling
+/// `AbortHandle::new` and `Abortable::new` manually.
+pub fn abortable<T>(future: T) -> (Abortable<T>, AbortHandle)
```

Comment:
Probably should be a method on `FutureExt`

The signature is similar to `StreamExt::split`:
```rust
fn split(self) -> (SplitSink<Self>, SplitStream<Self>) // method
```

Reply:
Additionally we could also add the method `future.abortable_via(abort_reg)`

Reply:
Personal preference but I think this makes more sense as a free function-- you wouldn't chain it like normal combinators since it doesn't return a `Future`.